phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -318,53 +322,123 @@ public int executeTool() {
+            ArrayList<String> immutableList,
+            long startWaitTime) {
+        int retryCount = 0;
+        long endWaitTime = System.currentTimeMillis();
+        long waitMore = getWaitMoreTime(endWaitTime, startWaitTime);
+        while  (retryCount < 5 && !waited) {
+            // If the table is immutable, we need to wait for clients to purge
+            // their caches of table metadata
 
+            LOGGER.info("waiting for more " + waitMore + " ms for client cache "
+                    + "to expire for immutable tables");
+            try {
+                startWaitTime = System.currentTimeMillis();
+                Thread.sleep(waitMore);
+                waited = true;
+            } catch (InterruptedException e) {
+                endWaitTime = System.currentTimeMillis();
+                waitMore = getWaitMoreTime(endWaitTime, startWaitTime);
+                retryCount++;
+                LOGGER.warning("Sleep before starting index rebuild is interrupted. "
+                        + "Attempting to sleep again! " + e.getMessage());
+            }
+        }
+        if (retryCount == 5) {
+            LOGGER.info("Sleep was interrupted multiple times for immutable tables, "
+                    + "we'll still go ahead and enable the table.");
+        }
+        try (Admin admin = queryServices.getAdmin()) {
+            for (String dataTableFullName: immutableList) {
+                HashSet<String> indexes = tablesAndIndexes.get(dataTableFullName);
+                enableTable(admin, dataTableFullName, indexes);
+            }
+        } catch (IOException | SQLException e) {
+            LOGGER.severe("Something went wrong while enabling immutable table " + e);
+        }
+    }
+
+    private long getWaitMoreTime(long endWaitTime, long startWaitTime) {
+        int waitTime = GLOBAL_INDEX_CHECKER_ENABLED_MAP_EXPIRATION_MIN+1;
+        if(test) {
+            return 1;
+        }
+        if(dryRun) {
+            return 0; //no wait
+        }
+        return (((waitTime) * 60000) - Math.abs(endWaitTime-startWaitTime));
+    }
+
+    private long executeToolForImmutableTables(ConnectionQueryServices queryServices,
+            ArrayList<String> immutableList) {
+        LOGGER.info("Started " + operation + " for immutable tables");
+        try (Admin admin = queryServices.getAdmin()) {
+            for (String dataTableFullName : immutableList) {
+                HashSet<String> indexes = tablesAndIndexes.get(dataTableFullName);
+                LOGGER.info("Executing " + operation + " of " + dataTableFullName
+                        + " (immutable)");
+                disableTable(admin, dataTableFullName, indexes);
+                modifyTable(admin, dataTableFullName, indexes);
+            }
+        } catch (IOException | SQLException e) {
+            LOGGER.severe("Something went wrong while disabling "
+                    + "or modifying immutable table " + e);
```

Comment:
I think we cannot simply ignore the failure here. We can downgrade the data table, remove it from the list and continue with the rest. If the downgrade fails too, then we can throw an exception.

Reply:
Thank you for the suggestion, kadir. I made the changes in the code.